### PR TITLE
Add ST25TB poller

### LINKED ISSUE #2072: ST25TB nfc tag support
### Description of the feature you're suggesting.

It would be nice to have support for ST25TB nfc tags. They're supported by [ST's RFAL](https://github.com/flipperdevices/flipperzero-firmware/blob/dev/lib/ST25RFAL002/include/rfal_st25tb.h) so it shouldn't be hard to add support at least in `furi_hal_nfc`. 
I'd need the support for a FAP I'm developing that reads, parses and writes back data to various types of tags so I don't need support in flipper gui etc.
I wouldn't mind using RFAL directly myself but if I understood everything correctly it isn't supported nor recommended.

### Anything else?

It's a NFC-B tag btw, not sure if it would need additional infra in furi hal :thinking:

### PR DESCRIPTION
**Pull request by augustozanellato**
Closes #2072

# What's new

- ST25TB poller
- ST25TB support in NFC app (only reading and saving)

# Verification 

- Scan st25tb tags in various sizes/versions
- Save them
- Load them back up

# Checklist (For Reviewer)

- [ ] PR has description of feature/bug or link to Confluence/Jira task
- [ ] Description contains actions to verify feature/bugfix
- [ ] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

**Comment by gornekich** [comment]
Hello @augustozanellato !
Wonderful work, thanks for the PR. We need some time to find these tags to test your PR and review the code.

Since you are the first one who add support for a new protocol for reworked NFC, could you leave some feedback? Was it hard to add new protocol? What part of the new NFC code was hard to understand?

**Comment by augustozanellato** [comment]
> Hello @augustozanellato ! Wonderful work, thanks for the PR. We need some time to find these tags to test your PR and review the code.

No problems with that, I submitted the pr with great advance so that you’ll have time to review and test it before refactored nfc gets merged :)

> Since you are the first one who add support for a new protocol for reworked NFC, could you leave some feedback?

Biggest issue imho was the lack of documentation, but that’s expected for a massive refactoring while it’s still in progress so don’t count that as feedback.
New api overall is nice, I’ve been following your branch commits for some time now 👀

> Was it hard to add new protocol?

Luckily not so much, thanks to the fact that there’s already a number of implemented protocols that can be taken as reference. 

> What part of the new NFC code was hard to understand?

The hardest part was probably understanding the interactions between nfc protocols and the rest of the firmware, eg the nfc app, but in the end it worked out great. 



Btw: if you need to source ST25TB tags you can get them either with the names found [here](https://github.com/flipperdevices/flipperzero-firmware/pull/3076/files#diff-21e74f781954a8df41d2335612a793f757538377e792cf93c37f9091ae70d578R22) or with the legacy naming (SRI4K, SRIX4K, SRI512, SRIX512 and SRT512), they’re the exact same tag (except maybe SRIX4K and SRIX512 because those have a “secret” authentication functionality, but that isn’t implemented in the polled).

**Comment by augustozanellato** [comment]
Just pushed a commit that shows the legacy naming in addition to the current one and also recognizes "X" tags such as SRIX4K and SRIX512.
Features are identical because only difference is support for the "french telecom anti clone protection" that's under nda but essentially is just a challenge/response auth protocol.

**Comment by augustozanellato** [comment]
Btw @gornekich I noticed a weird thing, maybe you'll be able to provide me with some insight.
Sometimes `st25tb_poller_async_activate` gets stuck with the following callstack:
image
Specifically it hangs inside `nfc_poller_trx_state_machine` at line 355 (`event = furi_hal_nfc_poller_wait_event(FURI_HAL_NFC_EVENT_WAIT_FOREVER);`), looks like it waits forever for an event that doesn't seem to come through.
Could it be a bug in my poller code or is it an issue in the state machine?
It doesn't happen 100% of the times so I'm not sure how to debug that.

EDIT: some clarifications: I tried to sniff the comms with my pm3 and the tag correctly responds to the command when the issue happens.

**Comment by gornekich** [comment]
@augustozanellato yes, I reproduce this issue sometimes.

There is TODO on 354 line addressed to this issue. Looks like sometimes NFC chip doesn't emit TxEnd interrupt. I saw this in errata and I need some time to make workaround.

Now I think that we shouldn't wait forever for this specific event. The fix will come soon

File: lib/nfc/protocols/st25tb/st25tb_poller.c
Comment (wosk) [inline lib/nfc/protocols/st25tb/st25tb_poller.c]:
```suggestion
    furi_assert(event.protocol == NfcProtocolInvalid);
```

File: lib/nfc/protocols/st25tb/st25tb_poller_i.c
Comment (wosk) [inline lib/nfc/protocols/st25tb/st25tb_poller_i.c]: doesn't matter

File: lib/nfc/protocols/st25tb/st25tb_poller.c
Comment (augustozanellato) [reply lib/nfc/protocols/st25tb/st25tb_poller.c]: Oof, thanks!

File: lib/nfc/protocols/st25tb/st25tb_poller_i.c
Comment (augustozanellato) [reply lib/nfc/protocols/st25tb/st25tb_poller_i.c]: Yeah that's useless, I left that behind while refactoring, thanks again :)

**Comment by augustozanellato** [comment]
> Now I think that we shouldn't wait forever for this specific event. The fix will come soon

Did the fix come through? I just rebased the pr on your branch's latest changes and it seems to be much more reliable.

**Comment by gornekich** [comment]
> > Now I think that we shouldn't wait forever for this specific event. The fix will come soon
> 
> 
> 
> Did the fix come through? I just rebased the pr on your branch's latest changes and it seems to be much more reliable.

Not yet. Will come in a few days.

**Comment by gornekich** [comment]
@augustozanellato We pushed fix for the mentioned issue. Please update your branch and let me know if there are still any problems.

Also I couldn't find any of these tags that I can get in reasonable time. May be you know some stores where we can order them? We really want flipper to support this protocol, but we have troubles with testing your PR on real tags.

**Comment by augustozanellato** [comment]
> @augustozanellato We pushed fix for the mentioned issue. Please update your branch and let me know if there are still any problems.

It seems to be much more reliable now, made 15 scans in a row without freezes.

> Also I couldn't find any of these tags that I can get in reasonable time. May be you know some stores where we can order them? We really want flipper to support this protocol, but we have troubles with testing your PR on real tags.

They're kinda difficult to buy sadly.
The only ones that are easy to buy are SRI512, the others usually have a large MOQ or are only sold as a bare die. SRI512 can be found for cheap on aliexpress, eg [here](https://aliexpress.com/item/1005001310193983.html) or [here](https://aliexpress.com/item/1005001310454677.html). SRT512 can be found on some rechargeable transport ticket systems, for examples some cities in France use those.

**Comment by skotopes** [comment]
Tested, works well. Waiting for @gornekich approval and merge

File: lib/nfc/protocols/st25tb/st25tb_poller_i.c
Comment (gornekich) [inline lib/nfc/protocols/st25tb/st25tb_poller_i.c]: Please, change to FURI_LOG_D()
If card is not present, there will be timeout error, which is normal situation and shouldn't be logged in Error log level

**Review by gornekich — CHANGES_REQUESTED** [review]
Hello @augustozanellato 
I reviewed the code, everything is good. Please fix the log level in one place and I will merge PR

File: lib/nfc/protocols/st25tb/st25tb_poller_i.c
Comment (augustozanellato) [reply lib/nfc/protocols/st25tb/st25tb_poller_i.c]: Fixed in 12533fb2aff0fbf54d692e2b61efd777cb00516a

**Comment by marchacio** [comment]
Hello everyone, is anyone trying to develop writing on these cards?
I'm not an expert on these things but I'm just starting to get my hands on it and I would need to be able to write data to an ST25TB tag.

I found some resources online that explain how writing and data validation works in ST25TB cards:

First link:
https://raw.githubusercontent.com/gentilkiwi/st25tb_kiemul/main/ST25TB_transport.pdf

Second link (this is a PDF, probably it will be downloaded automatically):
https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://www.st.com/resource/en/product_presentation/st25tb-product-presentation.pdf&ved=2ahUKEwiDrJmj7rOIAxUQ_rsIHV7JBAkQFnoECBUQAQ&usg=AOvVaw3qwnjG4LdJrl4HJuPqX79p


I was wondering if it was possible and if anyone was already working on implementing data writing. 
Thank you very much

**Comment by augustozanellato** [comment]
> Hello everyone, is anyone trying to develop writing on these cards? I'm not an expert on these things but I'm just starting to get my hands on it and I would need to be able to write data to an ST25TB tag.
> 
> I found some resources online that explain how writing and data validation works in ST25TB cards:
> 
> First link: https://raw.githubusercontent.com/gentilkiwi/st25tb_kiemul/main/ST25TB_transport.pdf
> 
> Second link (this is a PDF, probably it will be downloaded automatically): https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://www.st.com/resource/en/product_presentation/st25tb-product-presentation.pdf&ved=2ahUKEwiDrJmj7rOIAxUQ_rsIHV7JBAkQFnoECBUQAQ&usg=AOvVaw3qwnjG4LdJrl4HJuPqX79p
> 
> I was wondering if it was possible and if anyone was already working on implementing data writing. Thank you very much

As you can see here [1] and here [2] write support is already implemented in the poller, it’s just not exposed in the UI because it would probably just be a source of confusion in users willing to restore an old dump without knowing about the countdown counter blocks.

[1]: https://github.com/flipperdevices/flipperzero-firmware/blob/a122ee75f6711b449a1f172fc5c12d287e1997cc/lib/nfc/protocols/st25tb/st25tb_poller.h#L70
[2]: https://github.com/flipperdevices/flipperzero-firmware/blob/a122ee75f6711b449a1f172fc5c12d287e1997cc/lib/nfc/protocols/st25tb/st25tb_poller_sync.h#L12

**Comment by marchacio** [comment]
> > Hello everyone, is anyone trying to develop writing on these cards? I'm not an expert on these things but I'm just starting to get my hands on it and I would need to be able to write data to an ST25TB tag.
> > I found some resources online that explain how writing and data validation works in ST25TB cards:
> > First link: https://raw.githubusercontent.com/gentilkiwi/st25tb_kiemul/main/ST25TB_transport.pdf
> > Second link (this is a PDF, probably it will be downloaded automatically): https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://www.st.com/resource/en/product_presentation/st25tb-product-presentation.pdf&ved=2ahUKEwiDrJmj7rOIAxUQ_rsIHV7JBAkQFnoECBUQAQ&usg=AOvVaw3qwnjG4LdJrl4HJuPqX79p
> > I was wondering if it was possible and if anyone was already working on implementing data writing. Thank you very much
> 
> As you can see here [1](https://github.com/flipperdevices/flipperzero-firmware/blob/a122ee75f6711b449a1f172fc5c12d287e1997cc/lib/nfc/protocols/st25tb/st25tb_poller.h#L70) and here [2](https://github.com/flipperdevices/flipperzero-firmware/blob/a122ee75f6711b449a1f172fc5c12d287e1997cc/lib/nfc/protocols/st25tb/st25tb_poller_sync.h#L12) write support is already implemented in the poller, it’s just not exposed in the UI because it would probably just be a source of confusion in users willing to restore an old dump without knowing about the countdown counter blocks.


Oh, wow. you saved me days of work, thank you!
At this point I ask how can I use/implement these functions in my FZ machine?
Should I create my own custom firmware or are there other solutions?

Again, I just bought the FZ just two days ago and I still have to learn so much.
Thanks again for your help and patience

**Comment by augustozanellato** [comment]
> At this point I ask how can I use/implement these functions in my FZ machine?

Functions are already implemented into the firmware, specifically they are included in the SDK for 3rd party apps.

> Should I create my own custom firmware or are there other solutions?

A custom firmware isn't necessary because, as you can see, writing is implemented in the repo containing the official firmware.
You'd instead need to write a custom app (aka .fap), you can find a lot of tutorials for that online, beware to search for something recent that's using `ufbt` as the build tool.

**Comment by Nablap** [comment]
Bonjour, désolé de relancer le sujet plusieurs mois après, j'ai bien lus les commentaires et regarder un peut partout, et je vois que marchacio voulais pouvoir écrire et sûrement émuler ou cloner un tag nfc st25tb sur Flipper Zéro, et je suis dans le même cas pour un nfc st25tb512-at/sri512.. Je suis coincé quelqu'un a trouver la solution au problème ou un fichier supplémentaire à installer pour débloquer l'écriture et l'émulation... ? Merci d'avance 👌🏻

**Pull request merged**
